cvat/apps/annotation/validation.py

```diff
@@ -0,0 +1,214 @@
+        previous_frame = None
+        sequence_track_ids = set()
+
+        for frame in seq.frames:
+            reporter.frame = frame.name
+            reporter.count_frame(seq.name)
+
+            frame_track_ids = set()
+            duplicated_track_ids = set()
+            for bbox_index, bbox in enumerate(frame.bboxes, start=1):
+                reporter.bbox_index = bbox_index
+
+                if bbox.track_id in frame_track_ids:
+                    duplicated_track_ids.add(bbox.track_id)
+                frame_track_ids.add(bbox.track_id)
+                sequence_track_ids.add(bbox.track_id)
+
+                _validate_coordinates(bbox, reporter)
+                _validate_attributes(bbox, reporter)
+                _validate_class_immutability(bbox, class_id_by_track_id, reporter)
+                _validate_position_change(bbox, previous_frame, jump_threshold=2, reporter=reporter)
+
+            reporter.bbox_index = None
+            for track_id in duplicated_track_ids:
+                reporter.report_track_id_duplication(track_id)
+            previous_frame = frame
+
+        reporter.frame = None
+        _validate_track_id_are_consecutive(sequence_track_ids, reporter)
+    return reporter
+
+
+def _validate_attributes(bbox, reporter):
+    try:
+        int(bbox.track_id)
+    except (TypeError, ValueError):
+        reporter.report_invalid_track_id_value(bbox.track_id)
+
+    if bbox.class_id not in label_by_class_id:
+        reporter.report_invalid_class_id_value(bbox.class_id)
+
+
+def _validate_class_immutability(bbox, class_id_by_track_id, reporter):
+    prev_class_id = class_id_by_track_id.get(bbox.track_id, bbox.class_id)
+    if prev_class_id != bbox.class_id:
+        reporter.report_class_id_change(prev_class_id, bbox.class_id)
+    class_id_by_track_id[bbox.track_id] = bbox.class_id
+
+
+def _validate_track_id_are_consecutive(sequence_track_ids, reporter):
+    expected = range(1, len(sequence_track_ids) + 1)
+    expected = set(map(str, expected))
+    if sequence_track_ids != expected:
+        reporter.report_track_id_non_consecutive(sequence_track_ids)
+
+
+def _validate_position_change(bbox, previous_frame, jump_threshold, reporter):
+    previous_bbox = previous_frame and previous_frame.bbox_by_track_id.get(bbox.track_id)
+    if not (previous_bbox and is_valid_bbox(previous_bbox) and is_valid_bbox(bbox)):
+        return
+
+    bbox = BoundingBox.from_two_corners(bbox.xtl, bbox.ytl, bbox.xbr, bbox.ybr)
+    previous_bbox = BoundingBox.from_two_corners(previous_bbox.xtl, previous_bbox.ytl, previous_bbox.xbr, previous_bbox.ybr)
+    if bbox.almost_equals(previous_bbox, abs_tol=1e-5):
+        reporter.report_no_move()
+        return
+
+    embracing_bbox = bbox.get_embracing_box(previous_bbox)
+    width_jump = embracing_bbox.width / (bbox.width + previous_bbox.width)
+    height_jump = embracing_bbox.height / (bbox.height + previous_bbox.height)
+    if max(width_jump, height_jump) > jump_threshold:
```

Comment:
just curious: why is this something to report?

Reply:
It was proposed in ticket ( [No large jumps of bounding box coordinates in adjacent frames.](https://p.daedalean.ai/T4454) ) to implement such check.
I think it helps when there are multiple objects on the video and the annotator mixes up track ids.